Mathlib/Topology/Compactness/CompactlyGeneratedSpace.lean

```diff
@@ -0,0 +1,177 @@
+    CompactlyGeneratedSpace X := by
+  refine uCompactlyGeneratedSpace_of_continuous_maps fun f h' ↦
+    continuous_iff_isClosed.2 fun t ht ↦ h _ ?_
+  intro K _ _ _ g hg
+  exact ht.preimage (h' (CompHaus.of K) { toFun := g, continuous_toFun := hg })
+
+/-- A topological space `X` is compactly generated if a set `s` is open when `f ⁻¹' s` is
+open for every continuous map `f : K → X`, where `K` is compact Hausdorff. -/
+theorem compactlyGeneratedSpace_of_isOpen
+    (h : ∀ (s : Set X), (∀ {K : Type u} [TopologicalSpace K], [CompactSpace K] → [T2Space K] →
+      ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s)) → IsOpen s) :
+    CompactlyGeneratedSpace X := by
+  refine compactlyGeneratedSpace_of_isClosed fun s h' ↦ isOpen_compl_iff.1 <| h _ fun f hf ↦ ?_
+  rw [preimage_compl, isOpen_compl_iff]
+  exact h' f hf
+
+/-- In a compactly generated space `X`, a set `s` is closed when `f ⁻¹' s` is
+closed for every continuous map `f : K → X`, where `K` is compact Hausdorff. -/
+theorem CompactlyGeneratedSpace.isClosed' [CompactlyGeneratedSpace X] {s : Set X}
+    (hs : ∀ {K : Type u} [TopologicalSpace K], [CompactSpace K] → [T2Space K] →
+      ∀ (f : K → X), Continuous f → IsClosed (f ⁻¹' s)) : IsClosed s := by
+  rw [eq_compactlyGenerated (X := X), TopologicalSpace.compactlyGenerated, isClosed_coinduced,
+    isClosed_sigma_iff]
+  exact fun ⟨_, f⟩ ↦ hs f f.continuous
+
+/-- In a compactly generated space `X`, a set `s` is closed when `s ∩ K` is
+closed for every compact `K ⊆ X`. -/
+theorem CompactlyGeneratedSpace.isClosed [CompactlyGeneratedSpace X] {s : Set X}
+    (hs : ∀ ⦃K⦄, IsCompact K → IsClosed (s ∩ K)) : IsClosed s := by
+  refine isClosed' fun f hf ↦ ?_
+  rw [← Set.preimage_inter_range]
+  exact (hs (isCompact_range hf)).preimage hf
+
+/-- In a compactly generated space `X`, a set `s` is open when `f ⁻¹' s` is
+open for every continuous map `f : K → X`, where `K` is compact Hausdorff. -/
+theorem CompactlyGeneratedSpace.isOpen' [CompactlyGeneratedSpace X] {s : Set X}
+    (hs : ∀ {K : Type u} [TopologicalSpace K], [CompactSpace K] → [T2Space K] →
+      ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s)) : IsOpen s := by
+  rw [eq_compactlyGenerated (X := X), TopologicalSpace.compactlyGenerated, isOpen_coinduced,
+    isOpen_sigma_iff]
+  exact fun ⟨_, f⟩ ↦ hs f f.continuous
+
+/-- In a compactly generated space `X`, a set `s` is closed when `s ∩ K` is
+closed for every compact `K ⊆ X`. -/
+theorem CompactlyGeneratedSpace.isOpen [CompactlyGeneratedSpace X] {s : Set X}
+    (hs : ∀ ⦃K⦄, IsCompact K → IsOpen (s ∩ K)) : IsOpen s := by
+  refine isOpen' fun f hf ↦ ?_
+  rw [← Set.preimage_inter_range]
+  exact (hs (isCompact_range hf)).preimage hf
+
+/-- If the topology of `X` is coinduced by a continuous function whose domain is
+compactly generated, then so is `X`. -/
+theorem compactlyGeneratedSpace_of_coinduced {Y : Type u}
+    [tX : TopologicalSpace X] [tY : TopologicalSpace Y]
```

Comment:
I believe this theorem and the instances below are true for `UCompactlyGeneratedSpace` with any universe. Could you try to generalise?